### PR TITLE
[StackIR] Support source maps and DWARF with StackIR

### PR DESCRIPTION
**Pull request by kripken**
Helping #6509, this fixes debugging support for StackIR, which makes it more
possible to use StackIR in more places.

The fix is basically just to pass around some more state, and then to call the
parent with "please write debug info" at the correct times, mirroring the
similar calls in `BinaryenIRWriter`.

The relevant Emscripten tests pass, and the source map test modified
here produces identical output in StackIR and non-StackIR modes.

Remove `--new-wat-parser` in the test here - @tlively we don't need that
anymore do we?

### CODE REVIEW

File: test/passes/fannkuch3_manyopts_dwarf.bin.txt [hunk 1]
```diff
@@ -2303,7 +2303,7 @@ Contains section .debug_info (851 bytes)
 Contains section .debug_loc (1073 bytes)
 Contains section .debug_ranges (88 bytes)
 Contains section .debug_abbrev (333 bytes)
-Contains section .debug_line (2682 bytes)
+Contains section .debug_line (2642 bytes)
```
Comment (tlively) [inline test/passes/fannkuch3_manyopts_dwarf.bin.txt:2306]: Is this expected and accounted for?

**Review by tlively — APPROVED** [review]
Code LGTM, so LGTM overall if the test changes make sense.

File: test/passes/fannkuch3_manyopts_dwarf.bin.txt (hunk 1 above)
Comment (kripken) [reply test/passes/fannkuch3_manyopts_dwarf.bin.txt:2306]: I believe so, yes. StackIR saves a few bytes here and there, which allows encoded offsets to be a bit shorter in some cases. Any increase in size (in this or any other field) would have been very suspicious, however. A large decrease would also have been odd.

I did also run `llvm-dwarfdump --validate` on the output and it is equally valid as before this PR (which is, 99% valid but for one known pre-existing minor issue).

**Comment by kripken** [comment]
@tlively what about this question?

> Remove --new-wat-parser in the test here - @tlively we don't need that anymore do we?

(if we actually don't then I can simplify the test a little more I think)

**Comment by tlively** [comment]
Oh yes, sorry. `--new-wat-parser` can be removed.

**Comment by kripken** [comment]
Thanks, then I simplified the test now to have autogenerated CHECKs, since the checks are identical between StackIR and BinaryenIR (and without two parsers, there is just one output for BinaryenIR).

**Review by tlively — APPROVED** [review]

**Pull request merged**